# 數值計算與資料輸入
## 數學運算子
|運算子|描述|範例|
|---|---|---|
|+|加法|1+2=3|
|-|減法|2-1=1|
|*|乘法|2*3=6|
|/|浮點數除法|7/2=3.5|
|//|整數除法|7//2=3|
|%|餘數|7%3=1|
|**|次方|2**3=8|
### 整數 int
1. 數字前面不能加0 -> SuntaxError 語法錯誤

In [3]:
05

SyntaxError: leading zeros in decimal integer literals are not permitted; use an 0o prefix for octal integers (4223885067.py, line 1)

2. 除數不能是0 -> ZeroDivisionError 零除法錯誤 

In [4]:
5/0

ZeroDivisionError: division by zero

3. 數學優先順序
`()` > `**` > `正負` > `* / % //` > `+ -` > `=`

In [5]:
2 * (1 + 2) ** 2 - 2 ** 2 * 2 

10

3. str不能轉成int or float -> ValueError

In [7]:
int('123 Apple')

ValueError: invalid literal for int() with base 10: '123 Apple'

4. 沒辦法同時做到「將float轉成int」& 「str轉int」 -> ValueError

In [11]:
int("123.6")

ValueError: invalid literal for int() with base 10: '123.6'

5. True=1、False=0

In [13]:
print(True+2)
print(False + 5.0 )

3
5.0


6. 可以換行加減

In [10]:
1+2+\
    3\
-1

5

### int的範圍

In [15]:
a=10**100
print(a*a)

100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000


### 浮點數 float

In [34]:
print(float('-1.586544'))

print(float('1.0e4'))


-1.586544
10000.0


### 2、8、16進位
|進位|前輟|
|---|---|
|2|0b 0B|
|8|0o 0O|
|16|0x 0X|

*前輟：告訴Python解釋器如何解讀後面的數字

In [6]:
print(10) #10進位

print(0b10) #2進位

print(0o10) #8進位

print(0x10) #16進位

10
2
8
16


## 字串 Str
### 多行str
用`"""`包覆已分段文字。寫的方式跟備註很像，只是要給他一個變數裝起來而已。

In [33]:
a = """這是多行字串的寫法
AAA
BBB"""

print(a)

這是多行字串的寫法
AAA
BBB


# 看不懂這是啥
`>>> ''`

`    ''`

`>>> ""`

`    ''`

`>>> ''''''`

`    ''`

`>>> """"""`

`    ''`

`>>>``

In [51]:
print('')
print("")
print('''''')
print("""""")

In [52]:
''

''

In [53]:
""

''

In [54]:
''''''

''

In [55]:
""""""

''

1. 用`+=`增加str
2. 用`%s %()`插入str
3. **沒辦法**用`-=`減少str -> TypeError

In [19]:
i = ''
i += 'abcdefg'
i += 'hijklmnop'
print(i)

c="apple"
d="An %s a day keeps the doctor away." %(c)
print(d)

b = i
b -= 'hijklmnop'
print(b)

abcdefghijklmnop
An apple a day keeps the doctor away.


TypeError: unsupported operand type(s) for -=: 'str' and 'str'

4. **沒辦法**str+int -> TypeError

In [29]:
a="dfeisdhouethg"+12
print(a)

TypeError: can only concatenate str (not "int") to str

3. 用`+`連接str時中間不會有空格，用`,`連接str時會有空格(因為str(**object=b''**, encoding='utf-8', errors='strict'))

In [14]:
a="AAA"
b="BBB"
c="CCC"

print(a+b+c)
print(a,b,c)

AAABBBCCC
AAA BBB CCC


## 複合指定運算子
*python所有的數學運算子都能這樣寫
|Python|表示|
|---|---|
|x`+=`y|x=x+y|
|x`-=`y|x=x-y|
|x`/=`y|x=x/y|
|x`//=`y|x=x//y|
|x`%=`y|x=x%y|
|x`*=`y|x=x*y|
|x`**=`y|x=x**y|
|**JS**|**表示**|
|x=`++`y|x=y+1 y=y+1|
|x=y`++`|**x=y**|
|x=`--`y|x=y-1 =-1|
|x=y`--`|沒這種東西|

In [7]:
#x += y
x=5
y=6
x += y
print(x)

#x -= y
x=5
y=6
x -= y
print(x)

#x **= y
x=5
y=6
x **= y
print(x)

11
-1
15625


In [48]:
#money.py
money = 50000 
cell = int(input("請輸入手機金額:"))
money -= cell
print("剩餘款為:" + str(money)) 

剩餘款為:49850


## 特殊字元
|寫法|用途|
|---|---|
|`\\`|反斜線|
|`\'`|單引號|
|`\"`|雙引號|
|`\n`|換行|
|`\t`|固定間隔|

In [12]:
print("\\這是反斜線\\")
print("\'這是單引號\'")
print("\"這是雙引號\"")
print("\n這是換行\n")
print("\t這是固定間隔\t")

\這是反斜線\
'這是單引號'
"這是雙引號"

這是換行

	這是固定間隔	


## 輸入資料
用`input()`產生一個可以輸入資料的窗口，只能傳出str。用`.__class__`查詢變數的資料型態。

語法1：`input()` -> 沒有參數

語法2：`input(prompt)` -> 有參數(可以在括號裡面塞引述：str)

In [6]:
#語法1
print("Hello")
input()

#語法2
a = input("請輸入字串: ")
print(a,a.__class__ )

Hello
apple <class 'str'>


## Tuple
用`,`可同時輸出多個基礎運算值

In [3]:
#括號可省略
(1+2,8//3,8**3)

(3, 2, 512)

## 數學模組-math
python內建的工具模組之一(e.g.數學-math、統計-statistics、時間-time .etc)，如果想用要引入：

`import math`

### 四捨五入 & 無條件進位 & 無條件捨去
#### 四捨五入-round()
語法：`round(number, digits)`

digits 四捨五入取到小數點後第幾位

#### 無條件進位 & 無條件捨去-math.ceil() & math.floor()
1. 引入math：`import math`
2. 只能取到整數，如要取到小數點後要自己寫

##### 無條件進位

語法：`math.ceil(num)`

##### 無條件捨去

語法：`math.floor(num)`

In [7]:
import math

#四捨五入
a=1.1564846456
print(round(a,2))

#無條件進位
a=1.1564846456
print(math.ceil(a*100)/100)

#無條件捨去
a=1.1564846456
print(math.floor(a*100)/100)

1.16
1.16
1.15


### 圓周率 & log
`math.pi`：圓周率

`math.e`：log

In [8]:
print(math.pi)
print(math.e)

3.141592653589793
2.718281828459045


### 弧度&角度
`math.radians(num)`：角度(num)轉弧度

`math.degrees(num)`：弧度(num)轉角度

In [9]:
print(math.radians(30))
print(math.degrees(0.5235987755982988))

0.5235987755982988
29.999999999999996


### sin & cos & tan
x=弧度

`math.sin(x)`：sin(x)的角度

`math.cos(x)`：cos(x)的角度

`math.tan(x)`：tan(x)的角度

### 反的sin & cos & tan

`math.asin(x)`：用sin(x)的角度求sin(x)的弧度

`math.acos(x)`：用cos(x)的角度求cos(x)的弧度

`math.atan(x)`：用tan(x)的角度求tan(x)的弧度

In [59]:
#假設一個30度的直角三角形邊分別是1:math.sqrt(3):2
a=math.radians(30)
print(a)
print(f"{math.sin(a)} -> 1/2\n{math.cos(a)} -> math.sqrt(3)/2\n{math.tan(a)} -> math.sqrt1/(3)")
print(math.asin(0.49999999999999994))

0.5235987755982988
0.49999999999999994 -> 1/2
0.8660254037844387 -> math.sqrt(3)/2
0.5773502691896257 -> math.sqrt1/(3)
0.5235987755982988


## 統計模組-statistics
算鐘形曲線的，更專業要參考Numpy、SciPy。
### 平均值 & 中位數 & 眾數
`statistics.mean([數列])`：平均值

`statistics.median([數列])`：中位數

`statistics.mode([數列])`：眾數，數列中重複出現最多的數字

**偶個數時(數列有偶數個e.g.10個)**

`statistics.median_high([數列])`：較高的中位數

`statistics.median_low([數列])`：較低的中位數

In [1]:
import statistics

print(f"平均值：{statistics.mean([1,2,3,5,7,11,13,17,19])}") #平均值

a=[1,3,5,7,9,2,4,6,8,10]
print(f"中位數：{statistics.median(a)}") #中位數
print(f"較高的中位數：{statistics.median_high(a)}") #較高的中位數
print(f"較低的中位數：{statistics.median_low(a)}") #較低的中位數

print(f"眾數：{statistics.mode([1,2,2,3,3,3,4,4,4,5,6,7,8,5,5,5,5,6,8,8,7,7,7,7,1])}") 

平均值：8.666666666666666
中位數：5.5
較高的中位數：6
較低的中位數：5
眾數：5


### 母體標準差、變異數
`statistics.pstdev([數列])`：母體標準差=標準差

概念：全校500人，500人全部身高的標準差

定義：跟標準(平均)差多少 -> [離均差的平方/n]開根號=(平方合的平均-平均的平方)開根號

`statistics.pvariance([數列])`：母體變異數 [離均差的平方/n]

範例：

[1,3,4,4,3,9]

平均=(1+3+4+4+3+9)/6=4

標準差=`{[(1-4)**2+(3-4)**2+(4-4)**2+(4-4)**2+(3-4)**2+(9-4)**2]/6}**1/2=6**1/2`

變異數=6

### 樣本標準差、變異數
`statistics.stdev([數列])`：樣本標準差 跟樣本標準(平均)差多少 -> [離均差的平方/n-1]開根號

概念：全校500人，抽50人出來身高的標準差

`statistics.variance([數列])`：樣本變異數 [離均差的平方/n-1]

範例：

[1,3,4,4,3,9]

平均=(1+3+4+4+3+9)/6=4

標準差=`{[(1-4)**2+(3-4)**2+(4-4)**2+(4-4)**2+(3-4)**2+(9-4)**2]/5}**1/2=(36/5)**1/2`

變異數=36/5

至於為何樣本標準差是n-1，可以看這個：[從標準差除以n或除以n-1談起 丁春成](https://web.math.sinica.edu.tw/math_media/d291/29102.pdf)

In [2]:
arr = [1, 2, 3, 4, 5, 6, 7, 8, 9]
a = statistics.pstdev(arr)
b = statistics.pvariance(arr)
c = statistics.stdev(arr)
d = statistics.variance(arr)
print(a)
print(b)
print(c)
print(d)

2.581988897471611
6.666666666666667
2.7386127875258306
7.5


## 精準度模組-decimal
decimal分成2個模塊：

Decimal：精準計算

getcontext：建構環境

引入要特別寫：

In [3]:
from decimal import Decimal, getcontext

### decimal的算術語境
也就是getcontext()裡可以修改的地方

In [4]:
from decimal import *
getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

`prec=28` 精準度(precision)，預設28位。

`rounding=ROUND_HALF_EVEN` 四捨五入的方式：偶數捨入法。

1. 如果要捨入的數字（小數點後的位數）為5，則向最接近的偶數捨入。如果要捨入的數字為奇數，則向下捨入到最接近的偶數；如果要捨入的數字為偶數，則向上捨入到最接近的偶數。

2. 對於非5的數字，按照傳統的四捨五入規則進行捨入。

範例：

4.5捨入為4：因為5是奇數，所以向下捨入。

5.5捨入為6：因為5是奇數，所以向上捨入。

8.4捨入為8：按照傳統四捨五入規則，捨入到最接近的整數，因為4小於5。

`Emin=-999999, Emax=999999` 最小指數：-999999~最大指數：999999。

`capitals=1` 科學記號表示法用大寫E (e.g.1E+3) ;capitals=0表示用小寫e (e.g.1e+3)。

`clamp=0` 不截斷計算成果。

`flags=[]` 這是一個列表，用於存儲標誌或標記，通常用於標識特定的計算條件或狀態。但在這沒有提供具體的使用情境。

`traps=[InvalidOperation, DivisionByZero, Overflow]` 

可能發生的異常狀況：無效操作、除以0、溢出，會觸發上面這三個異常處理程序。

### 有個小限制
float轉換成二進位儲存時，會有點變化

In [4]:
a = 1.115
print(f"a：{Decimal(a)}") # Decimal(a) 的值會比較奇異一點

b = 1.125
print(f"b：{Decimal(b)}") # Decimal(b) 的值就沒變

a：1.1149999999999999911182158029987476766109466552734375
b：1.125


**解決方法：傳入str**

In [9]:
a = 1.115
print(f"a：{Decimal(str(a))}")

b = 1.125
print(f"b：{Decimal(str(b))}")

a：1.115
b：1.125


### 小數點後的精準四捨五入-quantize
#### round的限制：二進位儲存造成結果偏誤
round約整的方法並不是四捨五入，是偶數捨入法，也就是decimal的ROUND_HALF_EVEN，這個約整方法在round無法變更。

In [12]:
print(f"a：{round(1.115, 2)}")
# 這是因為在電腦內部，1.115 真正的值是 1.1149999999999999911182158029987476766109466552734375

print(f"b：{round(1.125, 2)}")
# 1.125 真正的值也是 1.125 

a：1.11
b：1.12


**解決方法：Decimal.quantize用str輸入** + ROUND_HALF_UP

總覺得有點不實際

我懶得在這改了，詳細&&正確的在總綱/數值計算

In [18]:
print(f"a：{Decimal('1.115').quantize(Decimal('0.00'), ROUND_HALF_UP)}")

print(f"b：{Decimal('1.125').quantize(Decimal('0.00'), ROUND_HALF_EVEN)}")

a：1.12
b：1.12


## 練習題
1. 以下的資料輸出哪一個是錯的?

(1) "I can add integers, like " + str(5) + " to strings."

(2) "Isaid"+("Hey"*2)+"Hey!"

(3) "The correct answer to this multiple choice exercise is answer number" + 2

(4) True + False

A：(3) str!+int

In [28]:
"The correct answer to this multiple choice exercise is answer number" + 2

TypeError: can only concatenate str (not "int") to str

2. 請問執行後的說明哪一個是對的?

a,b,c="pcschool",2016,3.41

print(b)

(1) 2016

(2) 3.41

(3) pcschool

A：(1)

In [30]:
a,b,c="pcschool",2016,3.41
print(b)

2016


3. 請問執行後的說明哪一個是對的?

test=("abc"+"!")*2

print(test)

(1) abc!2

(2) abc!abc!

(3) abc+!*2

A：(2)

In [32]:
test=("abc"+"!")*2
print(test)

abc!abc!


4. 請問計算後Z的內容哪一個是對的?

x=5

y=7

z=x+y+1

(1) 571

(2) 1

(3) 13

A：(3)

In [34]:
x=5
y=7 
z=x+y+1
print(z)

13


5. 請動手操作，並留意輸出結果

In [35]:
#num.py
a,b,c=4,2,5
d,f=3.25,5.5

print (a*b) 

print (a**b) 

print (a % 3) 

print (d + f) 

print (c//b) 

print (a/b)

8
16
1
8.75
2
2.0


6. 請問執行後的說明哪一個是對的?

g,h=9,3

print(g/h)

(1) 3 

(2) 0 

(3) 3.0

A：(3)

In [36]:
g,h=9,3
print(g/h)

3.0


7. 前提：讓使用者**輸入**被除數(整數)及除數(整數，不可以是零)，且程式會顯示兩數相除的商及餘數。

執行內容：

請輸入被除數(整數):45

請輸入除數(整數,不可以為0):4

商 11 餘數: 1

In [25]:
a=int(input("請輸入被除數(整數):"))
b=int(input("請輸入除數(整數,不可以為0):"))
print(f"商：{a//b} 餘數：{a%b}",sep="")

商：11 餘數：1


8. 前提：計算使用者輸入的2個任意數，程式會顯示2數相加的總和。

執行內容：

請輸入第一個數值:45.67

請輸入第二個數值:67.47

兩個數的和是xxx.xx

In [27]:
a=float(input("請輸入第一個數值:"))
b=float(input("請輸入第二個數值:"))
c=str(a+b)
print(f"兩者的和是{c}")

兩者的和是113.14


9. 前提：請使用者輸入一個任意數，程式會顯示此數的平方值及立方值

執行內容：

請輸入任意數:3

此數的平方是:9.0

此數的立方是:27.0

In [26]:
a=float(input("請輸入一個任意數："))
b=str(a**2)
c=str(a**3)
print(f"\n此數的平方是：{b}\n此數的立方是：{c}")


此數的平方是：9.0
此數的立方是：27.0


10. 前提：請以(複合指定運算子)設計程式，讓用者輸入三個任意數，程式會顯示3數相加的總和(float)

執行內容：

請輸入第一個數:87.9

請輸入第二個數:45

請輸入等三個數:87.5

三個數的總和為:220.4

In [28]:
a=float(input("請輸入第一個數："))
a+=float(input("請輸入第二個數："))
a+=float(input("請輸入第三個數："))
print(f"三個數的總和為：{a}")

三個數的總和為：220.4


11. 前提：讓使用者輸入梯形的上底、下底及高，程式會計算梯形的面積(上底加下底乘以高除以2)

執行內容：

請輸入梯形的上底(公分):5

請輸入梯形的下底(公分):10

請輸入梯形的高(公分):2

梯形的面積:15平方公分

In [29]:
a=float(input("請輸入梯形的上底(公分)："))
b=float(input("請輸入梯形的下底(公分)："))
c=float(input("請輸入梯形的高(公分)："))
print(f"梯形的面積：{round((a+b)*c//2)}平方公分",sep="")

梯形的面積：15平方公分


12. 前提：使用者輸入圓柱體的半徑及高，程式會計算圓柱體的體積。圓柱體體積的公式為「圓週率乘以半徑平方再乘以高」。

執行內容：

請輸入圓柱體的半徑(公分):10

請輸入圓柱體的高(公分):5

圓柱體的體積:xxxx立方公分

In [30]:
import math

a=str(int(input("請輸入圓柱體的半徑(公分)："))**2*int(input("請輸入圓柱體的高(公分)："))*math.pi)
print(f"圓柱體的體積：{a}立方公分")

圓柱體的體積：1570.7963267948965立方公分


13. 前提：讓使用者輸入直角三角形的對邊、斜邊，進而計算角度。

公式:sin(x) = 對邊 / 斜邊   x是radian(弧度)

請輸入對邊:2.5

請輸入斜邊:5

直角三角形的角度為:30degree

In [40]:
#方法1
import math

a=float(input("請輸入對邊："))/float(input("請輸入斜邊："))
if a == 0.5:
    print("直角三角形的角度為：30degree")
elif a == math.sqrt(2)/2: #我不會輸入根號2，所以算不出這個答案
    print("直角三角形的角度為：45degree")
elif a == math.sqrt(3)/2: #同上
    print("直角三角形的角度為：60degree")
else:
    print("輸入的質錯誤，請重新輸入")

直角三角形的角度為：30degree


In [31]:
#方法2
import math

result = round(math.degrees(math.asin(0.5)))
print(f"直角三角形的角度為：{result}degree" ,sep="")

直角三角形的角度為：30degree
